torch/csrc/jit/passes/frozen_conv_folding.cpp

```diff
@@ -0,0 +1,116 @@
+#include <torch/csrc/jit/passes/frozen_conv_folding.h>
+#include <c10/util/Exception.h>
+#include <torch/csrc/jit/ir/constants.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/fold_conv_bn.h>
+#include <torch/torch.h>
+
+namespace torch {
+namespace jit {
+
+using Tensor = at::Tensor;
+
+bool nonConstantParameters(Node* n) {
+  for (size_t i = 1; i < n->inputs().size(); i++) {
+    if (n->inputs().at(i)->node()->kind() != prim::Constant) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool supportedConvNode(Node* n) {
+  switch (n->kind()) {
+    case aten::conv1d:
+    case aten::conv2d:
+    case aten::conv3d:
+      return true;
+    case aten::_convolution: {
+      auto transposed_conv =
+          constant_as<bool>(n->namedInput("transposed")).value_or(true);
+      // dont handle transposed conv yet or not-constant transpose parameter
+      return !transposed_conv;
+    }
+    default:
+      return false;
+  }
+}
+
+void FoldFrozenConvBatchnorm(Block* b) {
```

Comment:
What is the difference between this pass and the one in `torch/csrc/jit/passes/fold_conv_bn.*`? Is it that this pass handles functional conv/bn sequences while the existing pass handles Conv-module/BN-module calls? If so, the name might be adjusted I think since it theoretically could occur in a non-frozen model too.

Reply:
The other pass handles modules. Basically, the other pass was written pre-freezing, I'm not sure there's any reason for it to exist anymore

Reply:
Originally it was written for graph-mode quantization, which is replaced with an FX version now anyway. I think we should get rid of it then (cc @jerryzh168 ).

Reply:
this is added by @kimishpatel I think.

there are still people using torchscript graph mode quantization I think so we can't remove it right now.

Reply:
Doesn't graph mode quantization using freezing though? why can't we run fusion post - freezing?

Reply:
process for quantization is: fusion - prepare - convert - freeze
fusion needs to happen before freeze because it affects the weight of conv which will be observed in prepare and quantized in convert

Reply:
As Jerry said the original conv+bn fusion cannot work post freezing. Plus since that one works at module levels we cannot quite make it work post freezing which inlines the graph. Ideally we add graph model fusion which works directly on graphs instead of modules, which this PR seems to be adding. Right now module based fusion is used in optimize_for_mobile. We can potentially move it after freezing. If this pass is for graph based fusion then is it necessary that it works only on frozen graph? Can it be made to work on non-frozen graphs too? That might be useful.

Reply:
There's not a great way to make this work on graphs that aren't frozen. You would have to remove the batchnorm weights from the parent module which isn't in general a valid transformation unless user-specified. That would also add a third conv + bn pass.